common/core/flow_engine.c

```diff
@@ -0,0 +1,265 @@
+  if (ENGINE_NUM_BUFFERS <= engine_buffer) {
+    return false;
+  }
+
+  /* Zero-ise the content of the queue and the queue struct */
+  memzero(engine_ctx[engine_buffer].buffer,
+          ENGINE_STACK_DEPTH * (sizeof(flow_step_t *)));
+  memzero(engine_ctx[engine_buffer].array_list_config, sizeof(array_list_t));
+
+  /* Initialize the array_list with the appropriate parameters */
+  array_list_initialize(engine_ctx[engine_buffer].array_list_config,
+                        engine_ctx[engine_buffer].buffer,
+                        sizeof(flow_step_t *),
+                        ENGINE_STACK_DEPTH * (sizeof(flow_step_t *)));
+
+  return true;
+}
+
+bool engine_add_next_flow_step(uint8_t engine_buffer,
+                               const flow_step_t *flow_step_ptr) {
+  bool result = false;
+
+  if (ENGINE_NUM_BUFFERS <= engine_buffer) {
+    return result;
+  }
+
+  /* We want to store the pointer to the flow step inputted from the argument
+   * and NOT the actual content, therefore we need to pass the double pointer to
+   * the API */
+  const flow_step_t **flow_step_dptr = &flow_step_ptr;
+  result = array_list_insert(engine_ctx[engine_buffer].array_list_config,
+                             flow_step_dptr);
+  return result;
+}
+
+bool engine_goto_next_flow_step(uint8_t engine_buffer) {
+  bool result = false;
+
+  if (ENGINE_NUM_BUFFERS <= engine_buffer) {
+    return result;
+  }
+
+  result = array_list_iterate_next(engine_ctx[engine_buffer].array_list_config);
+  return result;
+}
+
+bool engine_goto_prev_flow_step(uint8_t engine_buffer) {
+  bool result = false;
+
+  if (ENGINE_NUM_BUFFERS <= engine_buffer) {
+    return result;
+  }
+
+  result = array_list_iterate_back(engine_ctx[engine_buffer].array_list_config);
+  return result;
+}
+
+bool engine_get_current_flow_step(uint8_t engine_buffer,
+                                  flow_step_t **flow_step_dptr) {
+  bool result = false;
+
+  if (ENGINE_NUM_BUFFERS <= engine_buffer) {
+    return result;
+  }
+
+  result = array_list_get_element(engine_ctx[engine_buffer].array_list_config,
+                                  flow_step_dptr);
+  return result;
+}
+
+bool engine_delete_current_flow_step(uint8_t engine_buffer) {
+  bool result = false;
+
+  if (ENGINE_NUM_BUFFERS <= engine_buffer) {
+    return result;
+  }
+
+  result = array_list_delete_entry(engine_ctx[engine_buffer].array_list_config);
+  return result;
+}
+
+void engine_run(uint8_t engine_buffer) {
+  if (ENGINE_NUM_BUFFERS <= engine_buffer) {
+    return;
+  }
+
+  while (1) {
+    flow_step_t *current_flow = NULL;
+
+    /* Break if the stack or queue reached an empty state */
+    if ((false == engine_get_current_flow_step(engine_buffer, &current_flow)) ||
+        (NULL == current_flow)) {
+      break;
+    }
+
+    const evt_config_t *evt_config_ptr = current_flow->evt_cfg_ptr;
+    /* We must ensure that the `evt_config_ptr` is not NULL */
+    ASSERT(NULL != evt_config_ptr);
```

Comment:
EVT pointer being null should be checked on adding the pointer to the list in `engine_add_next_flow_step`

Reply:
added this check.. good catch